typing/uniqueness_analysis.ml

```diff
@@ -619,258 +654,309 @@ end = struct
+(** See [report_error] for explanation *)
+type boundary_reason =
+  | Paths_from_mod_class (* currently will never trigger *)
+  | Free_var_of_mod_class (* currently will never trigger *)
+  | Out_of_mod_class
 
-let _mark_shared paths occ =
-  let mark_one path = UF.singleton path (Shared occ) in
-  UF.pars (List.map (fun path -> mark_one path) paths)
+type boundary = {
+  cannot_force : Maybe_unique.cannot_force;
+  reason : boundary_reason;  (** which kind of boundary is being crossed? *)
+}
 
-let mark_maybe_unique paths unique_use occ =
-  let mark_one path =
-    UF.singleton path (Maybe_unique (Maybe_unique.singleton unique_use occ))
-  in
-  UF.pars (List.map (fun path -> mark_one path) paths)
+exception Boundary of boundary
 
-(* returns:
-   the updated value.
-   the new introduced bindings.
-   usage during the process
-*)
-let pattern_match_var ~loc id value =
-  match value with
-  | Match_single (paths, _, _) -> (value, Ienv.singleton id paths, UF.unused)
-  | Match_tuple values ->
-      let path = UF.Path.fresh_root_of_ident id in
-      let ienv = Ienv.singleton id [ path ] in
-      (* Mark all ps as seen, as we bind the tuple to a variable. *)
-      (* Can we make it aliases instead of used? Hard to do if we want
-         to preserve the tree-ness *)
-      ( Match_single ([ path ], loc, None),
-        ienv,
-        UF.seqs
-          (List.map
-             (fun (paths, loc', modes) ->
-               match modes with
-               | None -> UF.unused
-               | Some unique_use ->
-                   let occ = { Occurrence.loc = loc' } in
-                   mark_maybe_unique paths unique_use occ)
-             values) )
-
-(*
-handling pattern match of value against pat, returns ienv and uf.
-ienv is the new bindings introduced;
-uf is the usage caused by the pattern matching *)
-let rec pattern_match pat value =
+let force_shared_boundary t ~reason =
+  match Maybe_unique.force_shared t with
+  | Ok () -> ()
+  | Error cannot_force -> raise (Boundary { cannot_force; reason })
+
+type value_to_match =
+  | Match_tuple of Value.t list
+      (** The value being matched is a tuple; we treat it specially so matching
+  tuples against tuples merely create alias instead of uses; We need [Value.t]
+  instead of [Paths.t] because the tuple could be bound to a variable, in which
+    case all values in the tuple is considered used *)
+  | Match_single of Paths.t  (** The value being matched is not a tuple *)
+
+let conjuncts_pattern_match ?prologue l =
+  let exts, ufs = List.split l in
+  let ufs = match prologue with None -> ufs | Some uf -> uf :: ufs in
+  (Ienv.Extension.conjuncts exts, UF.seqs ufs)
+
+let rec pattern_match pat vtm =
+  match (vtm, pat.pat_desc) with
+  | _, Tpat_or (a, b, _) ->
+      let ext0, uf0 = pattern_match a vtm in
+      let ext1, uf1 = pattern_match b vtm in
+      (Ienv.Extension.disjunct ext0 ext1, UF.seq uf0 uf1)
+  | _, Tpat_lazy pat -> pattern_match pat vtm
```

Comment:
Yeah that sounds reasonable. For this reason I think `Match_single` also needs to contain a `Value.t` instead of just `Paths.t`; shouldn't be hard to change.